kaggle_environments/envs/werewolf/game/engine.py

```diff
@@ -0,0 +1,582 @@
+                    if player and player.alive:
+                        prompt = self.day_voting.get_voting_prompt(self.state, voter_id)
+                        self.state.push_event(
+                            description=prompt,
+                            event_name=EventName.VOTE_REQUEST,
+                            public=False,
+                            visible_to=[voter_id],
+                            visible_in_ui=False,
+                        )
+            return DetailedPhase.DAY_VOTING_AWAIT
+
+    @phase_handler(DetailedPhase.DAY_VOTING_CONCLUDE)
+    def _handle_day_voting_conclude(self, player_actions: Dict[PlayerID, Action]) -> DetailedPhase:
+        exiled_player_id = self.day_voting.get_elected()
+        if exiled_player_id:
+            exiled_player = self.state.get_player_by_id(exiled_player_id)
+            if exiled_player:
+                self.state.eliminate_player(exiled_player_id)
+
+                role = None
+                team = None
+                description = f'Player "{exiled_player_id}" is exiled by vote.'
+                if self._day_exile_reveal_level == RevealLevel.ROLE:
+                    role = exiled_player.role.name
+                    team = exiled_player.role.team
+                    description = (
+                        f'Player "{exiled_player_id}" in team {team} is exiled by vote. The player is a {role}.'
+                    )
+                elif self._day_exile_reveal_level == RevealLevel.TEAM:
+                    team = exiled_player.role.team
+                    description = f'Player "{exiled_player_id}" in team {team} is exiled by vote.'
+
+                data = DayExileElectedDataEntry(
+                    elected_player_id=exiled_player_id, elected_player_role_name=role, elected_player_team_name=team
+                )
+                self.state.push_event(description=description, event_name=EventName.ELIMINATION, public=True, data=data)
+        else:
+            self.state.push_event(
+                description="The vote resulted in no exile (e.g., a tie, no majority, or all abstained).",
+                event_name=EventName.VOTE_RESULT,
+                public=True,
+                data={"vote_type": "day_exile", "outcome": "no_exile", "reason": "tie_or_no_majority"},
+            )
+
+        self.day_voting.reset()
+        self.state.add_phase_divider(PhaseDivider.DAY_VOTE_END)
+        self.state.add_phase_divider(PhaseDivider.DAY_END)
+        return DetailedPhase.NIGHT_START
+
+    def _determine_and_log_winner(self):
+        # Check if a GAME_END entry already exists
+        game_end_event = self.state.get_event_by_name(EventName.GAME_END)
+        if game_end_event:
+            return  # Winner already logged for this day count
+
+        wolves = [p for p in self.state.alive_players() if p.role.team == Team.WEREWOLVES]
+        villagers = [p for p in self.state.alive_players() if p.role.team == Team.VILLAGERS]
+
+        if not wolves:
```

Comment:
It looks like we're missing a state for the case where a single player causes the game to end early, such as by timing out or delivering invalid responses. It looks like penalizing just the player causing the error wouldn't be too tricky to add here. Would it be a reasonable lift to integrate with the other error handling?